frontend/src/components/EditableLabelsDescriptionListGroup.tsx

```diff
@@ -1,223 +1,191 @@
+import React, { useState } from 'react';
+import { Label, LabelGroup, Alert, AlertVariant, Button } from '@patternfly/react-core';
+import DashboardDescriptionListGroup from './DashboardDescriptionListGroup';
 
-type EditableTextDescriptionListGroupProps = Partial<
-  Pick<DashboardDescriptionListGroupProps, 'title' | 'contentWhenEmpty'>
-> & {
+interface EditableLabelsProps {
   labels: string[];
-  saveEditedLabels: (labels: string[]) => Promise<unknown>;
-  allExistingKeys?: string[];
+  onLabelsChange: (labels: string[]) => Promise<void>;
   isArchive?: boolean;
-};
+  title?: string;
+  contentWhenEmpty?: string;
+  allExistingKeys: string[];
+}
 
-const EditableLabelsDescriptionListGroup: React.FC<EditableTextDescriptionListGroupProps> = ({
+export const EditableLabelsDescriptionListGroup: React.FC<EditableLabelsProps> = ({
   title = 'Labels',
   contentWhenEmpty = 'No labels',
   labels,
-  saveEditedLabels,
+  onLabelsChange,
   isArchive,
-  allExistingKeys = labels,
+  allExistingKeys,
 }) => {
-  const [isEditing, setIsEditing] = React.useState(false);
-  const [unsavedLabels, setUnsavedLabels] = React.useState(labels);
-  const [isSavingEdits, setIsSavingEdits] = React.useState(false);
+  const [isEditing, setIsEditing] = useState(false);
+  const [isSavingEdits, setIsSavingEdits] = useState(false);
+  const [unsavedLabels, setUnsavedLabels] = useState(labels);
+  const [labelErrors, setLabelErrors] = useState<{ [key: string]: string }>({});
 
-  const editUnsavedLabel = (newText: string, index: number) => {
-    if (isSavingEdits) {
-      return;
+  const validateLabel = (text: string, currentLabel?: string): string | null => {
+    if (currentLabel === text) {
+      return null;
+    }
+
+    const isDuplicate =
+      unsavedLabels.some((key) => key !== currentLabel && key === text) ||
+      allExistingKeys.some((key) => key !== currentLabel && key === text);
+
+    if (isDuplicate) {
+      return `"${text}" already exists. Use a unique name that doesn't match any existing key or property`;
+    }
+    if (text.length > 63) {
+      return "Label text can't exceed 63 characters";
     }
-    const copy = [...unsavedLabels];
-    copy[index] = newText;
-    setUnsavedLabels(copy);
+    return null;
   };
+
+  const handleEditComplete = (
+    _event: MouseEvent | KeyboardEvent,
+    newText: string,
+    currentLabel?: string,
+  ) => {
+    const error = validateLabel(newText, currentLabel);
+    if (error) {
+      setLabelErrors({ [newText]: error });
+      setUnsavedLabels((prev) => {
+        const filtered = prev.filter((label) => label !== currentLabel);
+        return [...filtered, newText];
+      });
+    } else if (newText) {
+      setUnsavedLabels((prev) => {
+        if (currentLabel) {
+          return [...prev, newText];
+        }
+        const filtered = prev.filter((label) => label !== currentLabel);
+        return [...filtered, newText];
+      });
+      setLabelErrors({});
+    }
+  };
```

Comment:
I'm not sure why the `setUnsavedLabels` behavior is different if there's an error vs if there isn't. This may resolve itself if you move the validation from here to render time, but I'm pretty sure `handleEditComplete` should update the unsaved labels state the same way no matter whether there is an error to show. I think this first call in your `if (error)` case is correct:

```ts
      setUnsavedLabels((prev) => {
        const filtered = prev.filter((label) => label !== currentLabel);
        return [...filtered, newText];
      });
```

Your other case is making it so any time edit focus is lost, if the user didn't enter duplicate label text, the old label is not filtered out on update and we end up with another duplicate:

https://github.com/user-attachments/assets/dfc40c98-19b2-4f21-9661-ea4637c4cca0